### PR TITLE
Update `WebSearch` to use a driver; add drivers for Google and DuckDuckGo

### LINKED ISSUE #792: Add DuckDuckGo Search tool
- [ x] I have read and agree to the [contributing guidelines](https://github.com/griptape-ai/griptape#contributing).

**Is your feature request related to a problem? Please describe.**
It can be difficult to use Google for search results if you don't have easy access to an API key. DuckDuckGo provides search without requiring a key.

**Describe the solution you'd like**
A DuckDuckGo search tool.

Here's an example of a tool, using the `duckduckgo_search` library:

```
from __future__ import annotations

import schema
from attr import define
from duckduckgo_search import DDGS
from griptape.artifacts import ErrorArtifact, ListArtifact, TextArtifact
from griptape.tools import BaseTool
from griptape.utils.decorators import activity
from schema import Literal, Schema


@define
class DuckDuckGoTool(BaseTool):
    @activity(
        config={
            "description": "Can be used to perform a text search on DuckDuckGo",
            "schema": Schema(
                {
                    Literal("keywords", description="The keywords to search for"): str,
                    schema.Optional(
                        "max_results",
                        default=None,
                        description="The maximum number of results to return",
                    ): int,
                    schema.Optional(
                        "region",
                        default="wt-wt",
                        description="The region to search in (wt-wt, us-en, uk-en, ru-ru, etc)",
                    ): str,
                    schema.Optional(
                        "safe_search",
                        default="moderate",
                        description="The safe search level to use (on, moderate, off)",
                    ): str,
                    schema.Optional(
                        "timelimit",
                        default=None,
                        description="The time limit for the search (d, w, m, y)",
                    ): str,
                    schema.Optional(
                        "backend",
                        default="api",
                        description="The backend to use for the search (api, html, lite)",
                    ): str,
                },
            ),
        }
    )
    def text(self, params: dict) -> ListArtifact | ErrorArtifact:
        keyword_value = params["values"].get("keywords")
        max_results = params["values"].get("max_results")
        region = params["values"].get("region")
        safe_search = params["values"].get("safe_search")
        timelimit = params["values"].get("timelimit")
        backend = params["values"].get("backend")

        # Initialize a dictionary with the keyword argument
        params = {}

        # Add parameters to the dictionary only if they are not None
        if max_results is not None:
            params["max_results"] = max_results
        if region is not None:
            params["region"] = region
        if safe_search is not None:
            params["safe_search"] = safe_search
        if timelimit is not None:
            params["timelimit"] = timelimit
        if backend is not None:
            params["backend"] = backend
        try:
            results = DDGS().text(keyword_value, **params)
            return ListArtifact([TextArtifact(result) for result in results])

        except Exception as e:
            return ErrorArtifact(str(e))

    @activity(
        config={
            "description": "Can be used to get instant answers from DuckDuckGo",
            "schema": Schema(
                {
                    Literal("keywords", description="The keywords to search for"): str,
                },
            ),
        }
    )
    def answers(self, params: dict) -> ListArtifact | ErrorArtifact:
        keyword_value = params["values"].get("keywords")

        try:
            results = DDGS().answers(keyword_value)
            return ListArtifact([TextArtifact(result) for result in results])

        except Exception as e:
            return ErrorArtifact(str(e))
```

### PR DESCRIPTION
**Pull request by vachillo**
Closes https://github.com/griptape-ai/griptape/issues/383
closes #792

### CODE REVIEW

File: griptape/drivers/prompt/huggingface_hub_prompt_driver.py [hunk 1]
```diff
@@ -80,6 +80,6 @@ def __prompt_stack_to_tokens(self, prompt_stack: PromptStack) -> list[int]:
         )
 
         if isinstance(tokens, list):
-            return tokens
+            return tokens  # type: ignore
```
Comment (collindutter) [inline griptape/drivers/prompt/huggingface_hub_prompt_driver.py:83]: What was the pyright error? Also seems unrelated to this PR?

File: griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py [hunk 2]
```diff
@@ -79,6 +79,6 @@ def __prompt_stack_to_tokens(self, prompt_stack: PromptStack) -> list[int]:
         )
 
         if isinstance(tokens, list):
-            return tokens
+            return tokens  # type: ignore
         else:
```
Comment (collindutter) [inline griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py:83]: What was the pyright error? Also seems unrelated to this PR?

File: griptape/drivers/web_search/duck_duck_go_web_search_driver.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import json
+from attrs import define, field
+from griptape.artifacts import TextArtifact, ListArtifact
+from griptape.drivers import BaseWebSearchDriver
+from duckduckgo_search import DDGS
+
+
+@define
+class DuckDuckGoWebSearchDriver(BaseWebSearchDriver):
+    client: DDGS = field(default=DDGS(), kw_only=True)
```
Comment (collindutter) [inline griptape/drivers/web_search/duck_duck_go_web_search_driver.py:10]: Use `import_optional_dependency` inside of `Factory`

File: griptape/drivers/web_search/duck_duck_go_web_search_driver.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import json
+from attrs import define, field
+from griptape.artifacts import TextArtifact, ListArtifact
+from griptape.drivers import BaseWebSearchDriver
+from duckduckgo_search import DDGS
```
Comment (collindutter) [inline griptape/drivers/web_search/duck_duck_go_web_search_driver.py:5]: Move into type checking

**Review by collindutter — COMMENTED** [review]
Great work. Just a minor fix on DDG and then we're good.

File: griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py (hunk 2 above)
Comment (vachillo) [reply griptape/drivers/prompt/huggingface_pipeline_prompt_driver.py]: temp fix for some errors i saw on precommit hooks, removed.

File: griptape/drivers/web_search/duck_duck_go_web_search_driver.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import json
```
Comment (collindutter) [inline griptape/drivers/web_search/duck_duck_go_web_search_driver.py:1]: Need: `from __future__ import annotations`

**Review by collindutter — DISMISSED** [review]

File: griptape/drivers/web_search/base_web_search_driver.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+
+@define
+class BaseWebSearchDriver(ABC):
+    results_count: int = field(default=5, kw_only=True)
+    language: str = field(default="en", kw_only=True)
+    country: str = field(default="us", kw_only=True)
+
+    @abstractmethod
+    def search(self, query: str, **kwargs) -> ListArtifact: ...
```
Comment (collindutter) [inline griptape/drivers/web_search/base_web_search_driver.py:14]: Wondering if we need `kwargs` here. I know we use this pattern in a couple other Drivers but we rarely seem to leverage it.

Comment (vachillo) [reply griptape/drivers/web_search/base_web_search_driver.py:14]: this was sort of future-proofing for passing in dynamic values

**Review by collindutter — DISMISSED** [review]

**Review by collindutter — DISMISSED** [review]

**Review by collindutter — APPROVED** [review]

**Review by collindutter — APPROVED** [review]
